### ISSUE TITLE
host groups

### ISSUE THREAD

**Issue by mhorne484** [cockpit-project/cockpit#2791]
I would love to see some organisation of hosts, could there be a feature to add groups?

for example, a web group that could contain all web servers? a mail group for mail servers and mailing list servers? workstations for local machines?

**Comment by stefwalter** [comment]
What would you do with such groups? Just visually separate them?

CC @petervo @andreasn

**Comment by michaelwemoto** [comment]
yes, we have a long list of servers (many, many virtual) that I have in
my host list.
I would love to group them so I can see servers separate from our local
workstations.
Then I can have sections for each category of server too. just an
organisational tweak :)

**Comment by stefwalter** [comment]
@michaelwemoto Is this something you're interested in working on? I've placed it on our ideas page:

https://github.com/cockpit-project/cockpit/wiki/Ideas

**Comment by michaelwemoto** [comment]
@stefwalter Unfortunately I'm not so much a dev as an admin, I wouldn't know where to begin. sorry I couldn't be of more help.

**Comment by stefwalter** [comment]
Makes sense. Hopefully someone will pick this up at some point. It may fit better once we have a more complete dashboard. It's at the top of the ideas page.

**Comment by roskogabor** [comment]
Hello guys, is there an update on this topic? I'm new cockpit user and have the same idea and if there is somebody who working on it I would like to help or if there isn't I will start to working on it.

**Comment by marusak** [comment]
I am afraid there is no update.
If I understand this correctly, this would go on the dashboard page, right? This page is being obsoleted and the hosts are going to be listed in a different place. We are currently making design for that, so we might need to take this issue into consideration as well.

**Comment by roskogabor** [comment]
@marusak How can I get more details about the changes?

**Comment by marusak** [comment]
@garrett is currently designing it.
Mockup can be seen here: https://garrett.github.io/cockpit-mockweb/navigation2/
To see the list of hosts, click on top left corner.

**Comment by garrett** [comment]
While at the face-to-face meeting during DevConf at the end of Jan, I wanted to have automatic grouping of some sort, but rethought the idea as the auto-grouping didn't make sense.

Since we're going to drop the graphs (as focused dashboard tooling would be better than anything we could provide), Cockpit would be able to scale to support additional hosts. Having some sort of grouping, as suggested in this issue, would help to make sense of a larger number of hosts.

(The alternative is to rename the hosts to be descriptive, such as `web-server.company.tld` instead of silly things like `banana` and `time-machine`. An admin using Cockpit might not have the ability to rename the hosts network-wide — and renaming hosts in existing environments may break many things.)

**TL;DR**: After seeing this issue* and reading the request, I think grouping makes sense and would like to work it into the dashboard mockup.

\* (thanks, @roskogabor, for asking about an update and then @marusak for mentioning me)

---

I think we'd have an ungrouped list of hosts by default (and don't show the group title). When the first group is made, it should then show the custom groups first, then the "ungrouped" group of hosts. 

Adding hosts to a group could be done in the edit host dialog, with a dropdown of groups with a "new group" option that would then show an input field for the new group name.

Additionally, when editing the hosts list with a toggle (as it's not a common occurance and the remove and edit buttons could otherwise be clicked too easily), we could have an edit group option exposed as well. Clicking it could then edit the group from a top-down instead of bottom-up approach: It would show a list of hosts and allow you to check them to move the hosts into that group.

_**In other words**_: Dynamic groups and groups are set in the host config settings (where colors are currently set). We could also possibly have group editing (once groups are active).

A mockup would probably be a bit easier to understand than my explanation above, so I'll try to work it in.

No promises on if it'd make it in or when, however. It would have to be done on a second pass. A first pass would include existing functionality (minus the graphs). We would ideally follow back up with another iteration or two to add status and/or group support.

But I do think this is would be a useful feature to have. Let's see what we can do. :wink:

**Comment by pwFoo** [comment]
Additional nice feature would be to trigger updates for host groups?

**Comment by KenCrandall** [comment]
Groups would also be a great way to have a quick list of hosts available for live migration in cockpit-machines.  That way you could choose from the hosts in the group or enter a URI if you need to do something specific/special

(I realize this is a year later, but I can think of no better place to put this)

**Comment by martinpitt** [comment]
We are actually moving away from managing multiple hosts. The dashboard is no more, there is a technical limit how many hosts you can connect to at the same time, and we even considered to entirely disable this functionality as it is insecure (we aren't actually going to, though). Moreover, this kind of grouping would be cockpit specific and only live in  a single browser's localstorage. I don't see an appropriate system API to do this.

As such, this is hard to implement, a fringe case, and low benefit, and hasn't made any progress in 7 years. We have plenty of other issues to work on.
